File: Java/src/main/java/com/gildedrose/BackstagePassesStrategy.java
package com.gildedrose;

public enum BackstagePassesStrategy implements QualityUpdateStrategy {
    INSTANCE;

    @Override
    public void applyTo(Item item) {
        if (item.sellIn <= 0) {
            item.quality = 0;
        } else if (item.sellIn >= 10) {
            item.quality = Math.min(50, item.quality+1);
        } else if (item.sellIn >= 5) {
            item.quality = Math.min(50, item.quality+2);
        } else {
            item.quality = Math.min(50, item.quality+3);
        }
        --item.sellIn;
    }
}

File: Kotlin/src/main/kotlin/com/platinumrose/mid/strategy/BackstagePassesStrategy.kt
package com.platinumrose.mid.strategy

import com.gildedrose.Item
import com.platinumrose.ItemConstant.Companion.REGULAR_ITEM_MAX_QUALITY
import com.platinumrose.ItemType.BACKSTAGE_PASSES


internal class BackstagePassesStrategy : UpdateQualityStrategy {

    override fun type(): com.platinumrose.ItemType {
        return BACKSTAGE_PASSES
    }

    override fun updateQuality(item: Item) {
        item.quality += 1
        if (item.sellIn < 11) {
            item.quality += 1
        }
        if (item.sellIn < 6) {
            item.quality += 1
        }
        item.sellIn -= 1
        if (item.sellIn < 0) {
            item.quality = 0
        }
    }

    override fun maxQuality(): Int {
        return REGULAR_ITEM_MAX_QUALITY
    }
}